ext/ffi/lib.rs

```diff
@@ -150,85 +157,99 @@ impl From<NativeType> for libffi::middle::Type {
+      "u32" => NativeType::U32,
+      "i32" => NativeType::I32,
+      "u64" => NativeType::U64,
+      "i64" => NativeType::I64,
+      "usize" => NativeType::USize,
+      "isize" => NativeType::ISize,
+      "f32" => NativeType::F32,
+      "f64" => NativeType::F64,
+      "string" => NativeType::String,
+      "buffer" => NativeType::Buffer,
+      _ => unimplemented!(),
+    }
+  }
+}
+
+enum NativeValue {
+  Void,
+  U8(u8),
+  I8(i8),
+  U16(u16),
+  I16(i16),
+  U32(u32),
+  I32(i32),
+  U64(u64),
+  I64(i64),
+  USize(usize),
+  ISize(isize),
+  F32(f32),
+  F64(f64),
+  String(*const i8),
+  Buffer(*const u8),
 }
 
 impl NativeValue {
-  fn new(native_type: NativeType, value: Value) -> Self {
+  fn from_value(native_type: NativeType, value: Value) -> Self {
     match native_type {
-      NativeType::Void => Self { void_value: () },
-      NativeType::U8 => Self {
-        u8_value: value_as_uint::<u8>(value),
-      },
-      NativeType::I8 => Self {
-        i8_value: value_as_int::<i8>(value),
-      },
-      NativeType::U16 => Self {
-        u16_value: value_as_uint::<u16>(value),
-      },
-      NativeType::I16 => Self {
-        i16_value: value_as_int::<i16>(value),
-      },
-      NativeType::U32 => Self {
-        u32_value: value_as_uint::<u32>(value),
-      },
-      NativeType::I32 => Self {
-        i32_value: value_as_int::<i32>(value),
-      },
-      NativeType::U64 => Self {
-        u64_value: value_as_uint::<u64>(value),
-      },
-      NativeType::I64 => Self {
-        i64_value: value_as_int::<i64>(value),
-      },
-      NativeType::USize => Self {
-        usize_value: value_as_uint::<usize>(value),
-      },
-      NativeType::ISize => Self {
-        isize_value: value_as_int::<isize>(value),
-      },
-      NativeType::F32 => Self {
-        f32_value: value_as_f32(value),
-      },
-      NativeType::F64 => Self {
-        f64_value: value_as_f64(value),
-      },
+      NativeType::Void => Self::Void,
+      NativeType::U8 => Self::U8(value_as_uint::<u8>(value)),
+      NativeType::I8 => Self::I8(value_as_int::<i8>(value)),
+      NativeType::U16 => Self::U16(value_as_uint::<u16>(value)),
+      NativeType::I16 => Self::I16(value_as_int::<i16>(value)),
+      NativeType::U32 => Self::U32(value_as_uint::<u32>(value)),
+      NativeType::I32 => Self::I32(value_as_int::<i32>(value)),
+      NativeType::U64 => Self::U64(value_as_uint::<u64>(value)),
+      NativeType::I64 => Self::I64(value_as_int::<i64>(value)),
+      NativeType::USize => Self::USize(value_as_uint::<usize>(value)),
+      NativeType::ISize => Self::ISize(value_as_int::<isize>(value)),
+      NativeType::F32 => Self::F32(value_as_f32(value)),
+      NativeType::F64 => Self::F64(value_as_f64(value)),
+      NativeType::String => Self::String(
+        CString::new(
+          value
+            .as_str()
+            .expect("Expected ffi arg value to be a string"),
+        )
+        .unwrap()
+        .into_raw(),
```

Comment:
This introduces a memory leak in most cases.
My suggestion: remove string support from the FFI interface.
To pass strings, the user can use TextEncoder/TextDecoder to create a Buffer.

Reply:
BTW: here's how `ffi-napi` does it, the string leaks too but you can use `finally` to release it
![image](https://user-images.githubusercontent.com/3526922/134245048-ec4c5154-3e1e-4dbc-b75f-f485b4df114d.png)